File: compiler_gym/service/runtime/benchmark_cache.py
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
import logging
from typing import Dict, Optional

import numpy as np

from compiler_gym.service.proto import Benchmark

MAX_SIZE_IN_BYTES = 512 * 104 * 1024

logger = logging.getLogger(__name__)


class BenchmarkCache:
    """An in-memory cache of Benchmark messages.

    This object caches Benchmark messages by URI. Once the cache reaches a
    predetermined size, benchmarks are evicted randomly until the capacity is
    reduced to 50%.
    """

    def __init__(
        self,
        max_size_in_bytes: int = MAX_SIZE_IN_BYTES,
        rng: Optional[np.random.Generator] = None,
    ):
        self._max_size_in_bytes = max_size_in_bytes
        self.rng = rng or np.random.default_rng()

        self._benchmarks: Dict[str, Benchmark] = {}
        self._size_in_bytes = 0

    def __getitem__(self, uri: str) -> Benchmark:
        """Get a benchmark by URI. Raises KeyError."""
        item = self._benchmarks.get(uri)
        if item is None:
            raise KeyError(uri)
        return item

    def __contains__(self, uri: str):
        """Whether URI is in cache."""
        return uri in self._benchmarks

    def __setitem__(self, uri: str, benchmark: Benchmark):
        """Add benchmark to cache."""
        # Remove any existing value to keep the cache size consistent.
        if uri in self._benchmarks:
            self._size_in_bytes -= self._benchmarks[uri].ByteSize()
            del self._benchmarks[uri]

        size = benchmark.ByteSize()
        if self.size_in_bytes + size > self.max_size_in_bytes:
            if size > self.max_size_in_bytes:
                logger.warning(
                    "Adding new benchmark with size %d bytes exceeds total "
                    "target cache size of %d bytes",
                    size,
                    self.max_size_in_bytes,
                )
            else:
                logger.debug(
                    "Adding new benchmark with size %d bytes "
                    "exceeds maximum size %d bytes, %d items",
                    size,
                    self.max_size_in_bytes,
                    self.size,
                )
            self.evict_to_capacity()

        self._benchmarks[uri] = benchmark
        self._size_in_bytes += size

        logger.debug(
            "Cached benchmark %s. Cache size = %d bytes, %d items",
            uri,
            self.size_in_bytes,
            self.size,
        )

    def evict_to_capacity(self, target_size_in_bytes: Optional[int] = None) -> None:
        """Evict benchmarks randomly to reduce the capacity below 50%."""
        evicted = 0
        target_size_in_bytes = (
            self.max_size_in_bytes // 2
            if target_size_in_bytes is None
            else target_size_in_bytes
        )

        while self.size and self.size_in_bytes > target_size_in_bytes:
            evicted += 1
            key = self.rng.choice(list(self._benchmarks.keys()))
            self._size_in_bytes -= self._benchmarks[key].ByteSize()
            del self._benchmarks[key]

        if evicted:
            logger.info(
                "Evicted %d benchmarks from cache. "
                "Benchmark cache size now %d bytes, %d items",
                evicted,
                self.size_in_bytes,
                self.size,
            )

    @property
    def size(self) -> int:
        """The number of items in the cache."""
        return len(self._benchmarks)

    @property
    def size_in_bytes(self) -> int:
        """The combined size of the elements in the cache, excluding the
        cache overhead.
        """
        return self._size_in_bytes

    @property
    def max_size_in_bytes(self) -> int:
        """The maximum size of the cache."""
        return self._max_size_in_bytes

    @max_size_in_bytes.setter
    def max_size_in_bytes(self, value: int) -> None:
        """Set a new maximum cache size."""
        self._max_size_in_bytes = value
        self.evict_to_capacity(target_size_in_bytes=value)

File: compiler_gym/service/runtime/BenchmarkCache.cc
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.
#include "compiler_gym/service/runtime/BenchmarkCache.h"

#include <glog/logging.h>

using grpc::Status;
using grpc::StatusCode;

namespace compiler_gym::runtime {

BenchmarkCache::BenchmarkCache(size_t maxSizeInBytes, std::optional<std::mt19937_64> rand)
    : rand_(rand.has_value() ? *rand : std::mt19937_64(std::random_device()())),
      maxSizeInBytes_(maxSizeInBytes),
      sizeInBytes_(0){};

const Benchmark* BenchmarkCache::get(const std::string& uri) const {
  auto it = benchmarks_.find(uri);
  if (it == benchmarks_.end()) {
    return nullptr;
  }

  return &it->second;
}

void BenchmarkCache::add(const Benchmark&& benchmark) {
  const size_t benchmarkSize = benchmark.ByteSizeLong();

  // Remove any existing value to keep the cache size consistent.
  const auto it = benchmarks_.find(benchmark.uri());
  if (it != benchmarks_.end()) {
    const size_t replacedSize = it->second.ByteSizeLong();
    benchmarks_.erase(it);
    sizeInBytes_ -= replacedSize;
  }

  if (sizeInBytes() + benchmarkSize > maxSizeInBytes()) {
    if (benchmarkSize > maxSizeInBytes()) {
      LOG(WARNING) << "Adding new benchmark with size " << benchmarkSize
                   << " bytes exceeds total target cache size of " << maxSizeInBytes() << " bytes";
    } else {
      VLOG(3) << "Adding new benchmark with size " << benchmarkSize
              << " bytes exceeds maximum size " << maxSizeInBytes() << " bytes, " << this->size()
              << " items";
    }
    evictToCapacity();
  }

  benchmarks_.insert({benchmark.uri(), std::move(benchmark)});
  sizeInBytes_ += benchmarkSize;

  VLOG(3) << "Cached benchmark " << benchmark.uri() << " (" << benchmarkSize
          << " bytes). Cache size = " << sizeInBytes() << " bytes, " << size() << " items";
}

void BenchmarkCache::evictToCapacity(std::optional<size_t> targetSize) {
  int evicted = 0;
  targetSize = targetSize.has_value() ? targetSize : maxSizeInBytes() / 2;

  while (size() && sizeInBytes() > targetSize) {
    // Select a benchmark randomly.
    std::uniform_int_distribution<size_t> distribution(0, benchmarks_.size() - 1);
    size_t index = distribution(rand_);
    auto iterator = std::next(std::begin(benchmarks_), index);

    // Evict the benchmark from the pool of loaded benchmarks.
    ++evicted;
    sizeInBytes_ -= iterator->second.ByteSizeLong();
    benchmarks_.erase(iterator);
  }

  if (evicted) {
    VLOG(2) << "Evicted " << evicted << " benchmarks from cache. Benchmark cache "
            << "size now " << sizeInBytes() << " bytes, " << benchmarks_.size() << " items";
  }
}

void BenchmarkCache::setMaxSizeInBytes(size_t maxSizeInBytes) {
  maxSizeInBytes_ = maxSizeInBytes;
  evictToCapacity(maxSizeInBytes);
}

}  // namespace compiler_gym::runtime
